matlab: init at R2024b

james-atkins opened the pull request:
Fixes #56887.

https://www.mathworks.com/products/matlab.html

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

james-atkins committed:
matlab: init at R2024b

arunoruto commented:
This looks awesome!
I tried to test build it via `nixpkgs-review`, but it fails telling me, that the attribute `gtk` is missing in `gnome2.gtk`.

ShamrockLee commented:
Unless multi-version support is intended, it should be moved to `pkgs/by-name/matlab/package.nix`.

arunoruto commented:
Result of `nixpkgs-review pr 356831` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>matlab</li>
  </ul>
</details>

ShamrockLee commented on pkgs/applications/science/math/matlab/generic.nix:
```diff
@@ -0,0 +1,105 @@
+  # The MATLAB installation isn't 100% deterministic due to how the classpath.txt file is generated.
+  # But mpm download does appear to be deterministic. So create a fixed output derivation from the
+  # downloaded files, and then run the installer.
+  matlab-download = stdenvNoCC.mkDerivation {
```
We could provide this as a fetcher (like `fetchFromMPM`) and use it to specify the `src` attribute of the `matlab` package.

ShamrockLee commented on pkgs/applications/science/math/matlab/fetcher.nix:
```diff
@@ -0,0 +1,42 @@
+  # The MATLAB installation isn't 100% deterministic due to how the classpath.txt file is generated.
+  # But mpm download does appear to be deterministic. So create a fixed output derivation from the
+  # downloaded files, and then run the installer.
+  outputHashAlgo = "sha256";
```
The `hash` attribute is meant to be algorithm-agnostic wherever possible. That's the purpose of the tree-wide migration from `sha256 =` to `hash =` in Nixpkgs.

```suggestion
  outputHashAlgo = lib.optionalString (hash == "") "sha256";
```

ShamrockLee commented on pkgs/applications/science/math/matlab/fetcher.nix:
```diff
@@ -0,0 +1,42 @@
+  outputHashMode = "recursive";
+  outputHash = hash;
+
+  passthru.matlab = {
+    inherit products;
+  };
```
```suggestion
```

ShamrockLee commented on pkgs/applications/science/math/matlab/fetcher.nix:
```diff
@@ -0,0 +1,42 @@
+  passthru.matlab = {
+    inherit products;
+  };
+}
```
```suggestion
})
```

ShamrockLee commented on pkgs/applications/science/math/matlab/mpm.nix:
We could consider exposing `mpm-bin` in `pkgs`. This would make debugging and overriding easier.

ShamrockLee reviewed (commented):
The fetchers and the main package are built assuming that packages from MPM must be fetched together, and Matlab must always be fetched together with other packages. This means the user will always need to re-fetch and rebuild `matlab` whenever they want additional software provided by `mpm`, such as Simulink.

Could Matlab and related packages like Simulink be packaged separately and made to depend on one another? That could save the user quite some bandwidth.

ShamrockLee commented on pkgs/applications/science/math/matlab/fetcher.nix:
```diff
@@ -0,0 +1,42 @@
+
+  installPhase = ''
+    # As mpm is bubblewrapped, it cannot write to $out directly
+    ${mpm}/bin/mpm download --release ${version} --platforms glnxa64 --destination matlab --products ${lib.escapeShellArgs products}
```
> mpm can't write to $out as it is bubblewrapped and /nix/store is mounted read-only (#239017). It needs to be bubblewrapped as the original downloaded executable unpacks another executable and then runs that.

How is "writing to `$out`" different from "writing to `./matlab` and rename it as `$out/matlab`"?

ShamrockLee replied:
Ah! I got it.

How about

```suggestion
    ${mpm}/bin/mpm download --release "$version" --platforms glnxa64 --destination matlab --products "$products"
```

2xsaiko commented on pkgs/applications/science/math/matlab/matlab.nix:
```diff
@@ -0,0 +1,161 @@
+
+  desktopItem = makeDesktopItem {
+    name = "matlab";
+    exec = "matlab -desktop -useStartupFolderPref";
```
```suggestion
    exec = "matlab -desktop -useStartupFolderPref %F";
```

dblsaiko reviewed (changes requested):
How do you use this for situations where what you have is

- the install ISO + install key
- *.lic files defining license servers that you want to add to the package for  system-wide use

Does it expose any programs in bin (matlab, mex, mlint for example, or a script to enter the FHS env)? I don't see where but I'm also not too familiar with buildFHSEnv.

For comparison, I also recently wrote a package which does all of the above. https://github.com/NixOS/nixpkgs/compare/master...2xsaiko:nixpkgs:push-nuzoruwspqxl

2xsaiko commented on pkgs/applications/science/math/matlab/matlab.nix:
```diff
@@ -0,0 +1,161 @@
+      "Math"
+      "Science"
+    ];
+    icon = "${matlab}/bin/glnxa64/cef_resources/matlab_icon.png";
```
Please use an icon name (e.g. "matlab") and not a path, since otherwise it won't allow themes overriding icons. Install/link the icon to $out/share/pixmaps.

arunoruto commented:
> idiomatic

Having an entry like `programs.matlab.license` would be ideal. Matlab supports the [`MLM_LICENSE_FILE` and `LM_LICENSE_FILE`](https://www.mathworks.com/help/install/ug/network-license-files.html) variables and can be set to a path in the nix store.

james-atkins committed:
Support matlab binaries both in bin and bin/glnxa64

james-atkins committed:
Purge MathWorks service host

james-atkins committed:
Bump mpm to 2024.4

james-atkins committed:
Fix FOD hashes for mpm 2024.4

mpm 2024.4 now bundles mpm with the downloaded files. Remove this so the
FOD hashes are the same as prior mpm versions.

james-atkins committed:
Move to by-name

ShamrockLee commented on pkgs/by-name/ma/matlab/package.nix:
```diff
@@ -0,0 +1,242 @@
+
+  unwrapped =
+    let
+      unpack =
+        if self.src ? "matlab-download" then
+          # Downloaded using fetchFromMPM: already unpacked
+          { dontUnpack = true; }
+        else
+          # Assume this is an ISO image from requireFile: need to unpack
+          {
+            nativeBuildInputs = [ libarchive ];
+            unpackCmd = "bsdtar -vxf $curSrc";
+            sourceRoot = ".";
+          };
```
The unpack phase doesn't apply in-place to `src`; the latter is read-only in the build sandbox. Are you trying to use the files unpacked into the build directory or use `$src` directly?

ShamrockLee replied:
> Thanks! I've pushed some commits which I think address this. Could you have a look?

The unpacking logic is now closed to most other programs.

Still, I guess one of the motivations behind the previous implementation is to avoid unnecessary copying of the large binaries downloaded by matlab-package-manager.

To achieve that, I would suggest keeping `dontUnpack = true` and deligating all the unpacking work to the derivation assigned to `src`. The Matlab install command in the install phase can then use `$src` directly.

ShamrockLee replied:
As for the ISO unpacking logic, the assumption and implementation seem hard to discover if we just buried them in the source code.

How about we drop them and document the overriding in the Nixpkgs Manual with an example of using the ISO image instead? If it is necessary to provide such support in the implementation, we should make it explicit by guarding it with an overridable switch (e.g., `self.isISOSource`)

james-atkins committed:
Fix installation from ISO

!!! This is not tested as I do not have access to an ISO !!!

Unpack ISO contents to "source".
If installing from mpm downloaded files, link them to "source" too.

james-atkins committed:
Refine unpack phase

If src is the fetcher, use the standard unpack phase.
Else, try to unpack the ISO using bsdtar.
